src/messaging/ExchangeDelegate.h

```diff
@@ -91,7 +91,16 @@ class DLL_EXPORT ExchangeDelegate
     /**
      * @brief
      *   This function is the protocol callback to invoke when the associated
-     *   exchange context is being closed
+     *   exchange context is being closed.
+     *
+     *   If the exchange was in a state where it was expecting a message to be
+     *   sent due to an earlier WillSendMessage call or because the exchange has
+     *   just been created as an initiator, the consumer is holding a reference
+     *   to the exchange and it's the consumer's responsibility to call
+     *   Release() on the exchange at some point.  The usual way this happens is
```

Comment:
This design is not robust and potentially error prone. It forces the user to override `OnExchangeClosing` and track exchange state, then decides whether to call `Release`. This is too complex for somebody doesn't understand exchange well.

I don't think the exchange is recoverable once `OnExchangeClosing` is triggered, so we should do the clean up no matter in which condition. Once `OnExchangeClosing` is raised, the user automatically lose the reference.

Reply:
We could in fact do that, and I considered that: the "waiting for send" state just drop the ref internally, and notifies the delegate to drop its ref.

The problem is that then we would end up with dangling pointers in all the delegates that can be in the "waiting to send" state that don't implement OnExchangeClosing.  Which is not implemented by _anyone_ right now except for timed handler and the RemoveFabric hack, afaict.

We could try changing all of our delegates to implement OnExchangeClosing and drop their refs, etc, etc.  This is absolutely doable, and the result is likely to be saner once it's all done, but it would be a lot more risky change than clearly codifying the current contract, which consumers already follow.